py/stream.c

```diff
@@ -47,14 +47,25 @@
 
 STATIC mp_obj_t stream_readall(mp_obj_t self_in);
 
+mp_obj_t mp_stream_op_supported(mp_obj_t self_in, mp_stream_op_t op) {
+    struct _mp_obj_base_t *o = (struct _mp_obj_base_t *)self_in;
+    if (o->type->stream_p == NULL) {
+        // CPython: io.UnsupportedOperation, OSError subclass
+        nlr_raise(mp_obj_new_exception_msg(&mp_type_OSError, "stream object required"));
+    } else if (op == MP_STREAM_OP_READ && o->type->stream_p->read == NULL) {
+        nlr_raise(mp_obj_new_exception_msg(&mp_type_OSError, "object with stream.read required"));
```

Comment:
I'd say having distinct message is too costly. Are the code size picture before and after the patch? The expected behavior is that these errors should happen in correctly-written application.  We may consider doing about them something when we receive >1 reports that someone was confused by a single generic message.